regression in master-experimental gateway frontends with older daemons

Given a build described here

https://github.com/scaytrase/docker-experimental-syntax-issue

I run `make build` and receive the following error 

```
 > [1/2] RUN cat /home/www-data/result:
------
failed to solve with frontend dockerfile.v0: failed to build LLB: executor failed running [/bin/sh -c cat /home/www-data/result]: unable to find user www-data: no matching entries in passwd fi
le
make: *** [build-usage] Error 1
```

(could be any command, even `RUN echo ""` fails the same way)

Rolling back syntax to  `20191217-secrets`. Any syntax later than this fails with the same error

https://github.com/scaytrase/docker-experimental-syntax-issue/blob/fixed/intermediate/Dockerfile#L1

Makes the build pass and work perfect.

--- a/frontend/gateway/grpcclient/client.go
+++ b/frontend/gateway/grpcclient/client.go
@@ -105,22 +105,39 @@ func (c *grpcClient) Run(ctx context.Context, f client.BuildFunc) (retError erro
 					Metadata: res.Metadata,
 				}
 				if res.Refs != nil {
-					m := map[string]*pb.Ref{}
-					for k, r := range res.Refs {
-						id, err := convertRef(r)
-						if err != nil {
-							retError = err
-							continue
+					if c.caps.Supports(pb.CapProtoRefArray) == nil {
+						m := map[string]*pb.Ref{}
+						for k, r := range res.Refs {
+							id, err := convertRef(r)
+							if err != nil {
+								retError = err
+								continue
+							}
+							m[k] = pb.NewRef(id)
 						}
-						m[k] = pb.NewRef(id)
+						pbRes.Result = &pb.Result_Refs{Refs: &pb.RefMap{Refs: m}}
+					} else {
+						m := map[string]string{}
+						for k, r := range res.Refs {
+							id, err := convertRef(r)
+							if err != nil {
+								retError = err
+								continue
+							}
+							m[k] = id
+						}
+						pbRes.Result = &pb.Result_RefsDeprecated{RefsDeprecated: &pb.RefMapDeprecated{Refs: m}}
 					}
-					pbRes.Result = &pb.Result_Refs{Refs: &pb.RefMap{Refs: m}}
 				} else {
 					id, err := convertRef(res.Ref)
 					if err != nil {
 						retError = err
 					} else {
-						pbRes.Result = &pb.Result_Ref{Ref: pb.NewRef(id)}
+						if c.caps.Supports(pb.CapProtoRefArray) == nil {
+							pbRes.Result = &pb.Result_Ref{Ref: pb.NewRef(id)}
+						} else {
+							pbRes.Result = &pb.Result_RefDeprecated{RefDeprecated: id}
+						}
 					}
 				}
 				if retError == nil {

--- a/frontend/gateway/pb/caps.go
+++ b/frontend/gateway/pb/caps.go
@@ -19,6 +19,11 @@ const (
 	CapReadDir                 apicaps.CapID = "readdir"
 	CapStatFile                apicaps.CapID = "statfile"
 	CapImportCaches            apicaps.CapID = "importcaches"
+
+	// CapProtoRefArray is a capability to return arrays of refs instead of single
+	// refs. This capability is only for the wire format change and shouldn't be
+	// used in frontends for feature detection.
+	CapProtoRefArray apicaps.CapID = "proto.refarray"
 )
 
 func init() {
@@ -89,6 +94,13 @@ func init() {
 	Caps.Init(apicaps.Cap{
 		ID:      CapImportCaches,
 		Name:    "import caches",
+		Enabled: true,
+		Status:  apicaps.CapStatusExperimental,
+	})
+
+	Caps.Init(apicaps.Cap{
+		ID:      CapProtoRefArray,
+		Name:    "wire format ref arrays",
 		Enabled: true,
 		Status:  apicaps.CapStatusExperimental,
 	})

--- a/solver/pb/caps.go
+++ b/solver/pb/caps.go
@@ -288,5 +288,4 @@ func init() {
 		Enabled: true,
 		Status:  apicaps.CapStatusExperimental,
 	})
-
 }

